spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java

```diff
@@ -302,14 +306,31 @@ private UnionReader(List<ValueReader<?>> readers) {
 
     @Override
     public InternalRow read(Decoder decoder, Object reuse) throws IOException {
-      InternalRow struct = new GenericInternalRow(readers.length);
+      // first we need to filter out NULL alternative if it exists in the union schema
+      int nullIndex = -1;
+      List<Schema> alts = schema.getTypes();
+      for (int i = 0; i < alts.size(); i++) {
+        Schema alt = alts.get(i);
+        if (Objects.equals(alt.getType(), Schema.Type.NULL)) {
+          nullIndex = i;
+          break;
+        }
+      }
+      InternalRow struct = new GenericInternalRow(nullIndex >= 0 ? alts.size() - 1 : alts.size());
+      for (int i = 0; i < struct.numFields(); i += 1) {
+        struct.setNullAt(i);
+      }
```

Comment:
for union `[string, null, int, long]` wouldn't that still lead to `[null, 1, null]` if, for example, the `int` option is set?

Reply:
> The previous implementation doesn't handle return values correctly when there is a null element in the union, like [null, int, string], it returned a struct data of [null, 1, null instead of just [1, null], this PR fixes it.

what is the desired behavior handling unions with null element?